src/Form/HearingPostRequestType.php

```diff
@@ -49,11 +48,41 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'label' => $this->translator->trans('Title', [], 'case'),
                 'help' => $this->translator->trans('Choose a title for the hearing post', [], 'case'),
             ])
-            ->add('template', ChoiceType::class, [
+            ->add('template', EntityType::class, [
+                'class' => MailTemplate::class,
                 'placeholder' => $this->translator->trans('Choose a template', [], 'case'),
                 'label' => $this->translator->trans('Mail template', [], 'case'),
                 'choices' => $templateChoices,
             ])
+        ;
+
+        $formModifier = function (FormInterface $form, MailTemplate $mailTemplate = null) use ($builder) {
+            $form->add('customData', MailTemplateCustomDataType::class, [
+                'label' => false,
+                'template' => $mailTemplate,
+                'data' => $builder->getData()->getCustomData(),
+            ]);
+        };
+
+        $builder->addEventListener(
+            FormEvents::PRE_SET_DATA,
+            function (FormEvent $event) use ($formModifier) {
+                $data = $event->getData();
+
+                $formModifier($event->getForm(), $data->getTemplate());
+            }
+        );
+
+        $builder->get('template')->addEventListener(
+            FormEvents::POST_SUBMIT,
+            function (FormEvent $event) use ($formModifier) {
+                $template = $event->getForm()->getData();
+
+                $formModifier($event->getForm()->getParent(), $template);
+            }
+        );
```

Comment:
Can we put this into a shared super-class or a trait with a “addTemplate(FormBuilderInterface)” (with custom fields) function to remove code duplication of handling of templates and custom fields?

Reply:
Added a `TemplateFormTrait`.